### PR TITLE
Changes to notify account updates in MySqlAccountService to consumers

### PR DESCRIPTION
**Pull request by Arun-LinkedIn**
Contains following changes:
1. Notify account updates in MySqlAccountService to consumers
2. Add version field in Container
3. Move selectInactiveContainersAndMarkInZk to AbstractAccountService

### CODE REVIEW

File: ambry-account/src/main/java/com/github/ambry/account/AbstractAccountService.java [hunk 1]
```diff
@@ -220,5 +218,57 @@ protected void notifyAccountUpdateConsumers(AccountInfoMap newAccountInfoMap, Ac
...
+        } catch (AccountServiceException ase) {
+          updateException = ase;
+          retry++;
+          Thread.sleep(config.retryDelayMs);
+        }
+      }
+      if (updateException != null) {
+        logger.error("Failed to mark containers INACTIVE in set : {}  after {} retries", inactiveContainerCandidateSet,
+            config.maxRetryCountOnUpdateFailure, updateException);
+        accountServiceMetrics.accountUpdatesToZkErrorCount.inc();
```
Comment (lightningrob) [inline ambry-account/src/main/java/com/github/ambry/account/AbstractAccountService.java:269]: This seems like the wrong metric (or maybe needs name change).

File: ambry-account/src/main/java/com/github/ambry/account/AccountUtils.java [hunk 2]
```diff
@@ -70,4 +74,51 @@ static boolean hasDuplicateAccountIdOrName(Collection<Account> accounts) {
     deprecatedContainers.addAll(accountService.getContainersByStatus(Container.ContainerStatus.INACTIVE));
     return deprecatedContainers;
   }
+
+  /**
+   * Selects {@link Container}s to be marked as INACTIVE. Check the valid data size of each DELETE_IN_PROGRESS container
+   * from {@link StatsSnapshot} and select the ones with zero data size to be marked as INACTIVE.
+   */
```
Comment (lightningrob) [inline ambry-account/src/main/java/com/github/ambry/account/AccountUtils.java:81]: Minor: add @return line.

**Review by lightningrob — CHANGES_REQUESTED** [review]
Good refactoring.  Approved after minor changes.

File: ambry-account/src/main/java/com/github/ambry/account/AccountInfoMap.java [hunk 3]
```diff
@@ -200,16 +199,21 @@ boolean hasConflictingAccount(Collection<Account> accountsToSet) {
...
-
-      // TODO: Once we have versioning, check that version for existing container (being updated) matches.
+      // if the container already exists, check that the snapshot version matches the expected value.
+      Container containerInMap = getContainerByNameForAccount(container.getParentAccountId(), container.getName());
+      if (containerInMap != null && container.getSnapshotVersion() != containerInMap.getSnapshotVersion()) {
+        logger.error(
+            "Container to update in AccountId {} (containerId={} containerName={}) has an unexpected snapshot version in store (expected={}, encountered={})",
+            parentAccountId, container.getId(), container.getName(), container.getSnapshotVersion(),
+            containerInMap.getSnapshotVersion());
+        return true;
```
Comment (jsjtzyy) [inline ambry-account/src/main/java/com/github/ambry/account/AccountInfoMap.java:209]: Is there a unit test for this?

File: ambry-account/src/main/java/com/github/ambry/account/MySqlAccountService.java [hunk 4]
```diff
@@ -401,11 +392,14 @@ private void updateContainersWithMySqlStore(short accountId, Collection<Containe
...
-        // new container added (insert into container table)
+        // new container added. Insert record into container table.
         mySqlAccountStore.addContainer(containerToUpdate);
       } else {
         if (!containerInCache.equals(containerToUpdate)) {
-          // existing container modified (update container table)
+          // Existing container modified. Increase the version of container and update container record
+          containerToUpdate =
+              new ContainerBuilder(containerToUpdate).setSnapshotVersion(containerToUpdate.getSnapshotVersion() + 1)
+                  .build();
```
Comment (jsjtzyy) [inline ambry-account/src/main/java/com/github/ambry/account/MySqlAccountService.java:402]: Same here, do we have unit test to verify the updated version?

**Review by jsjtzyy — APPROVED** [review]
LGTM after a few minor comments are addressed

**Review by jsjtzyy — APPROVED** [review]

**Review by lightningrob — APPROVED** [review]

File: ambry-api/src/main/java/com/github/ambry/account/Container.java [hunk 5]
```diff
@@ -459,7 +464,8 @@ private Container(JSONObject metadata, short parentAccountId) throws JSONExcepti
   Container(short id, String name, ContainerStatus status, String description, boolean encrypted,
       boolean previouslyEncrypted, boolean cacheable, boolean mediaScanDisabled, String replicationPolicy,
       boolean ttlRequired, boolean securePathRequired, Set<String> contentTypeWhitelistForFilenamesOnDownload,
-      boolean backupEnabled, short parentAccountId, long deleteTriggerTime, long lastModifiedTime) {
+      boolean backupEnabled, short parentAccountId, long deleteTriggerTime, long lastModifiedTime,
+      int snapshotVersion) {
```
[the remark this answers is not in the document]
Comment (Arun-LinkedIn) [reply ambry-api/src/main/java/com/github/ambry/account/Container.java:468]: I think Yingyi added isSameContainer() recently to only compare user editable fields in Container. @jsjtzyy

**Comment by lightningrob** [comment]
A few tests in MySqlAccountServiceIntegrationTest are failing.

File: ambry-api/src/main/java/com/github/ambry/account/Container.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (justinlin-linkedin) [reply ambry-api/src/main/java/com/github/ambry/account/Container.java:468]: Please add snapshotVersion in equals method.

**Review by justinlin-linkedin — APPROVED** [review]

File: ambry-account/src/main/java/com/github/ambry/account/AbstractAccountService.java [hunk 6]
```diff
@@ -220,5 +219,57 @@ protected void notifyAccountUpdateConsumers(AccountInfoMap newAccountInfoMap, Ac
       logger.debug("HelixAccountService is updated with 0 updated account");
     }
   }
+
+  /**
+   * Selects {@link Container}s to be marked as INACTIVE and marked in underlying account store.
+   */
+  public void selectInactiveContainersAndMarkInStore(StatsSnapshot statsSnapshot) {
```
[the remark this answers is not in the document]
Comment (SophieGuo410) [reply ambry-account/src/main/java/com/github/ambry/account/AbstractAccountService.java:226]: @justinlin-linkedin , I'm still working on the orca job implementation(it's in a different MP ambry-ecl-tools). Currently it's working on my local. But still need to tested on EI and Prod. I'll remove the legacy code once I've done with the orca job implementation.

**Pull request merged**